nemo/lightning/megatron_parallel.py

```diff
@@ -0,0 +1,975 @@
+        ...
+
+    def on_megatron_reduce_microbatches_end(self, *args, **kwargs) -> None:
+        ...
+
+    def on_megatron_log_step_end(self, *args, **kwargs) -> None:
+        ...
+
+    def on_megatron_step_end(self, *args, **kwargs) -> None:
+        ...
+
+
+ReductionT = TypeVar("ReductionT")
+
+
+class MegatronLossReduction(nn.Module, Generic[DataT, ReductionT]):
+    def __init__(self) -> None:
+        super(MegatronLossReduction, self).__init__()
+        self.batch = None
+        self.register_forward_pre_hook(self._pre_forward_hook)
+
+    def setup(self, batch) -> None:
+        self.batch = batch
+
+    def _pre_forward_hook(self, module, x):
+        return (self.batch,) + x
+
+    def forward(self, batch: DataT, forward_out: torch.Tensor) -> Tuple[torch.Tensor, ReductionT]:
+        raise NotImplementedError("Must be implemented by subclass.")
+
+    @abc.abstractmethod
+    def reduce(self, losses_reduced_per_micro_batch: Sequence[ReductionT]) -> torch.Tensor:
+        raise NotImplementedError("Must be implemented by subclass.")
+
+
+@runtime_checkable
+class MegatronCallbackProtocol(Protocol):
+    def __call__(self, tensor: torch.Tensor) -> Tuple[torch.Tensor, Dict[str, torch.Tensor]]:
+        ...
+
+
+@runtime_checkable
+class MegatronStepProtocol(Protocol):
+    def __call__(
+        self,
+        *,
+        forward_step_func,
+        data_iterator: Union[Iterator, List[Iterator]],
+        model: Union[torch.nn.Module, List[torch.nn.Module]],
+        num_microbatches: int,
+        seq_length: int,
+        micro_batch_size: int,
+        decoder_seq_length: Optional[int] = None,
+        forward_only: bool = False,
+        collect_non_loss_data: bool = False,
+    ) -> list:
+        ...
+
+
+def _calc_number_of_params(model: List[nn.Module]) -> int:
+    assert isinstance(model, list)
+
+    return sum([sum([p.nelement() for p in model_module.parameters()]) for model_module in model])
+
+
+def is_list_of_iterators(var) -> bool:
+    if not isinstance(var, list):
+        return False
+
+    return all(isinstance(item, collections.abc.Iterator) for item in var)
+
+
+def _make_data_iterator_list(model, data_iterator: Iterator) -> List[Iterator]:
+    """Convert data iterator into form expected by Megatron.
+
+    With interleaved pipeline parallelism, Megatron expects a
+    list of one data iterator per model chunk. Each model
+    chunk independently gets data from its data iterator, so
+    we need to interact with the data iterator multiple times
+    for each microbatch step. Instead of incorporating this
+    logic into the data loader, we cache the iterator's output
+    to the first model chunk and reuse it in the other model
+    chunks.
+    """
+    if not isinstance(model, list) or len(model) == 1:
+        return data_iterator  # TODO @tmoon: Remove
+        # TODO @tmoon: Use once available in Megatron-LM
+        # return DataIteratorList([data_iterator])
```

Comment:
Can we clean this up?

Reply:
Let's clean this up everywhere soon, since this is copied from elsewhere in NeMo.